### PR TITLE
[SYCL] Store stream buffers in the scheduler

### PR DESCRIPTION
**Pull request by againull**
Stream buffers need to be alive after submitting a kernel because it is
executed by the scheduler asynchronously. For this reason currently
stream buffers are stored in an associated stream object. This stream
object is passed to the handler and then forwarded further to a command
group to keep stream buffers alive for the scheduler.

But there is a problem with this approach. A command group cannot be
destroyed while stream buffers (which are accessed in this command
group) are alive. Stream buffers are destroyed only if the stream
is destroyed. Stream object is destroyed only if command group is
destroyed. So, there is a loop dependency. Which results in memory leaks.

Solution is to store stream buffers in the scheduler for each stream.
With this approach resources are released properly.

### CODE REVIEW

File: sycl/source/detail/scheduler/scheduler.hpp [hunk 1]
```diff
@@ -715,6 +715,38 @@ class Scheduler {
 
   friend class Command;
   friend class DispatchHostTask;
+
+  class StreamBuffers {
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/scheduler.hpp:719]: We need at least some info on this class.

Comment (s-kanaev) [inline sycl/source/detail/scheduler/scheduler.hpp:719]: It's all public. Why is it class then? Suggest switching to struct here.

File: sycl/source/detail/stream_impl.cpp [hunk 2]
```diff
@@ -17,27 +17,55 @@ namespace detail {
...
+// Method to provide an atomic access to the offset in the global stream
+// buffer and offset in the flush buffer
+GlobalOffsetAccessorT stream_impl::accessGlobalOffset(handler &CGH) {
+  auto OffsetSubBuf = buffer<char, 1>(
+      detail::Scheduler::getInstance().StreamBuffersPool.find(this)->second.Buf,
+      id<1>(0), range<1>(OffsetSize));
+  auto ReinterpretedBuf = OffsetSubBuf.reinterpret<unsigned, 1>(range<1>(2));
+  return ReinterpretedBuf.get_access<cl::sycl::access::mode::atomic>(
+      CGH, range<1>(2), id<1>(0));
+}
```
Comment (s-kanaev) [inline sycl/source/detail/stream_impl.cpp:57]: Are these functions guarded to be available at host only?

File: sycl/test/basic_tests/stream/release_resources_test.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+// RUN: env SYCL_PI_TRACE=2 %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
+// RUN: env SYCL_PI_TRACE=2 %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
+
+//==----------------------- release_resources_test.cpp ---------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
Comment (s-kanaev) [inline sycl/test/basic_tests/stream/release_resources_test.cpp:12]: I think this part isn't needed here.

**Review by s-kanaev — COMMENTED** [review]
Looks good overall.
A couple of comments only.
Some LIT tests failed also.

File: sycl/source/detail/stream_impl.cpp (hunk 2 above)
Comment (againull) [reply sycl/source/detail/stream_impl.cpp:57]: Yes, correct

File: sycl/source/detail/scheduler/scheduler.hpp (hunk 1 above)
Comment (againull) [reply sycl/source/detail/scheduler/scheduler.hpp]: Added info in the style of this file

Comment (againull) [reply sycl/source/detail/scheduler/scheduler.hpp]: Agree, switched to structure

File: sycl/test/basic_tests/stream/release_resources_test.cpp (hunk 3 above)
Comment (againull) [reply sycl/test/basic_tests/stream/release_resources_test.cpp]: Removed

**Review by againull — COMMENTED** [review]
Also fixed CI fails. Problem is that scheduler is a static object. And looks like some runtimes (for example, cuda, cpu) are unloaded before destructor of the scheduler is called. When buffers are deleted we at least call wait() for events in the scheduler, so this is a problem if device runtime is unloaded.  
That is why I added changes to remove buffers for a stream object as soon as stream is flushed. We can do this because flushing is a blocking operation at this moment but this is going to be changed in the future.

**Review by s-kanaev — DISMISSED** [review]

**Comment by bader** [comment]
@againull, please, take a look at CI regressions.

**Comment by againull** [comment]
> @againull, please, take a look at CI regressions.

Alexey, I am quite sure it is caused by https://github.com/intel/llvm/pull/2409

**Comment by bader** [comment]
> > @againull, please, take a look at CI regressions.
> 
> @bader , I am quite sure it is caused by #2409

@againull, why do you think so? There are no CI failures in this PR pre-commit.
@yanfeng3721, could you confirm that #2409 introduced this regression?

**Comment by againull** [comment]
> > > @againull, please, take a look at CI regressions.
> > 
> > 
> > @bader , I am quite sure it is caused by #2409
> 
> @againull, why do you think so? There are no CI failures in this PR pre-commit.
> @yanfeng3721, could you confirm that #2409 introduced this regression?

This test FAIL: SYCL::image_accessor_readsampler.cpp failed in precommit for #2409, but adding some new changes to #2409 triggered new testing and it passed, I believe it is a flaky failure.

**Comment by bader** [comment]
@againull, @yanfeng3721, please, either revert the #2409 caused the regression or temporarily disable failing test.
Please, keep CI system sane.

**Comment by yanfeng3721** [comment]
Hi , I  have create a PR: https://github.com/intel/llvm/pull/2422  to disable the flaky test temporarily .

**Pull request merged**